File: python/src/lichen/rpl/dis.py
# SPDX-License-Identifier: GPL-3.0-or-later
# SPDX-FileCopyrightText: The contributors to the LICHEN project
"""Authenticated DIS solicitation handling (RFC 6550 Sections 6.7.9 and 8.3)."""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from ipaddress import IPv6Address

from lichen.rpl.messages import DIS, RplError, RplOption, RplOptionType
from lichen.rpl.trickle import TrickleTimer

_VERSION_PREDICATE = 0x80
_INSTANCE_PREDICATE = 0x40
_DODAG_PREDICATE = 0x20
_SOLICITED_INFORMATION_LENGTH = 19


class DisAction(Enum):
    """Action selected after a link-authenticated, replay-admitted DIS."""

    IGNORE = "ignore"
    RESET_TRICKLE = "reset_trickle"
    UNICAST_DIO_WITH_CONFIGURATION = "unicast_dio_with_configuration"


@dataclass(frozen=True)
class SolicitedInformation:
    """Predicates carried by the RFC 6550 Solicited Information option."""

    rpl_instance_id: int
    flags: int
    dodag_id: IPv6Address
    version: int

    @classmethod
    def from_option(cls, option: RplOption) -> SolicitedInformation:
        if option.type != RplOptionType.SOLICITED_INFORMATION:
            raise RplError("not a Solicited Information option")
        if len(option.data) != _SOLICITED_INFORMATION_LENGTH:
            raise RplError(
                "Solicited Information option data must be exactly 19 bytes"
            )
        return cls(
            rpl_instance_id=option.data[0],
            flags=option.data[1],
            dodag_id=IPv6Address(option.data[2:18]),
            version=option.data[18],
        )

    def matches(self, rpl_instance_id: int, dodag_id: IPv6Address, version: int) -> bool:
        """Return true when every enabled predicate matches local state."""
        return (
            not self.flags & _INSTANCE_PREDICATE
            or self.rpl_instance_id == rpl_instance_id
        ) and (
            not self.flags & _DODAG_PREDICATE or self.dodag_id == dodag_id
        ) and (not self.flags & _VERSION_PREDICATE or self.version == version)


def handle_authenticated_dis(
    dis_wire: bytes,
    *,
    destination_is_multicast: bool,
    rpl_instance_id: int,
    dodag_id: IPv6Address,
    version: int,
    trickle: TrickleTimer,
    now_ms: int,
) -> DisAction:
    """Handle one DIS after link signature and replay admission.

    Multicast solicitations that match reset the DIO Trickle timer. Unicast
    solicitations that match request an immediate unicast DIO containing the
    DODAG Configuration option; they do not reset Trickle. The caller owns DIO
    construction/transmission and MUST invoke this function only with an
    authenticated, replay-admitted DIS.
    """
    if type(dis_wire) is not bytes:
        raise TypeError("DIS wire image must be exact bytes")
    if type(destination_is_multicast) is not bool:
        raise TypeError("destination_is_multicast must be bool")
    if type(rpl_instance_id) is not int or not 0 <= rpl_instance_id <= 0xFF:
        raise ValueError("RPLInstanceID must be an exact u8")
    if type(dodag_id) is not IPv6Address:
        raise TypeError("DODAGID must be an IPv6Address")
    if type(version) is not int or not 0 <= version <= 0xFF:
        raise ValueError("DODAGVersionNumber must be an exact u8")
    if type(trickle) is not TrickleTimer:
        raise TypeError("trickle must be an exact TrickleTimer")

    dis = DIS.from_bytes(dis_wire)
    solicited: SolicitedInformation | None = None
    for option in dis.options:
        if option.type != RplOptionType.SOLICITED_INFORMATION:
            continue
        if solicited is not None:
            raise RplError("duplicate Solicited Information option")
        solicited = SolicitedInformation.from_option(option)

    matches = solicited is None or solicited.matches(rpl_instance_id, dodag_id, version)
    if not matches:
        return DisAction.IGNORE
    if not destination_is_multicast:
        return DisAction.UNICAST_DIO_WITH_CONFIGURATION

    trickle.reset(now_ms)
    return DisAction.RESET_TRICKLE

File: rust/lichen-rpl/src/dis.rs
//! Authenticated DIS solicitation handling (RFC 6550 Sections 6.7.9 and 8.3).

use crate::message::{Dis, OptionIter, RplError};
use crate::trickle::TrickleTimer;

const OPT_SOLICITED_INFORMATION: u8 = 7;
const VERSION_PREDICATE: u8 = 0x80;
const INSTANCE_PREDICATE: u8 = 0x40;
const DODAG_PREDICATE: u8 = 0x20;
const SOLICITED_INFORMATION_LEN: usize = 19;

/// Action selected after a link-authenticated, replay-admitted DIS.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum DisAction {
    Ignore,
    ResetTrickle,
    UnicastDioWithConfiguration,
}

/// Local DODAG state used by Solicited Information predicates.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct DisContext {
    pub rpl_instance_id: u8,
    pub dodag_id: [u8; 16],
    pub version: u8,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
struct SolicitedInformation {
    rpl_instance_id: u8,
    flags: u8,
    dodag_id: [u8; 16],
    version: u8,
}

impl SolicitedInformation {
    fn from_data(data: &[u8]) -> Result<Self, RplError> {
        if data.len() != SOLICITED_INFORMATION_LEN {
            return Err(RplError::InvalidOption);
        }
        Ok(Self {
            rpl_instance_id: data[0],
            flags: data[1],
            dodag_id: data[2..18]
                .try_into()
                .map_err(|_| RplError::InvalidOption)?,
            version: data[18],
        })
    }

    fn matches(self, context: DisContext) -> bool {
        (self.flags & INSTANCE_PREDICATE == 0 || self.rpl_instance_id == context.rpl_instance_id)
            && (self.flags & DODAG_PREDICATE == 0 || self.dodag_id == context.dodag_id)
            && (self.flags & VERSION_PREDICATE == 0 || self.version == context.version)
    }
}

/// Handle one DIS after the caller has verified its link signature and replay tuple.
///
/// A matching multicast DIS resets the DIO Trickle timer. A matching unicast
/// DIS requests an immediate unicast DIO with a DODAG Configuration option and
/// deliberately leaves Trickle unchanged. DIO construction and transmission
/// remain the caller's responsibility.
pub fn handle_authenticated_dis(
    wire: &[u8],
    destination_is_multicast: bool,
    context: DisContext,
    trickle: &mut TrickleTimer,
    now_ms: u64,
    rand_offset: u32,
) -> Result<DisAction, RplError> {
    Dis::from_bytes(wire)?;
    let mut solicited = None;
    for option in OptionIter::new(Dis::options_tail(wire)) {
        let option = option?;
        if option.opt_type != OPT_SOLICITED_INFORMATION {
            continue;
        }
        if solicited.is_some() {
            return Err(RplError::InvalidOption);
        }
        solicited = Some(SolicitedInformation::from_data(option.data)?);
    }

    if solicited.is_some_and(|information| !information.matches(context)) {
        return Ok(DisAction::Ignore);
    }
    if !destination_is_multicast {
        return Ok(DisAction::UnicastDioWithConfiguration);
    }

    trickle.reset(now_ms, rand_offset);
    Ok(DisAction::ResetTrickle)
}

#[cfg(test)]
mod tests {
    use super::*;

    const CONTEXT: DisContext = DisContext {
        rpl_instance_id: 7,
        dodag_id: [0x22; 16],
        version: 9,
    };

    fn timer() -> TrickleTimer {
        let mut timer = TrickleTimer::new(4_000, 8, 10);
        timer.start(0, 0);
        timer.fire_transmit();
        timer.expire(4_000, 0);
        timer.heard_consistent();
        timer
    }

    fn solicited(flags: u8, instance: u8, dodag_id: [u8; 16], version: u8) -> [u8; 21] {
        let mut option = [0u8; 21];
        option[0] = OPT_SOLICITED_INFORMATION;
        option[1] = SOLICITED_INFORMATION_LEN as u8;
        option[2] = instance;
        option[3] = flags;
        option[4..20].copy_from_slice(&dodag_id);
        option[20] = version;
        option
    }

    fn dis_with_options(options: &[u8]) -> std::vec::Vec<u8> {
        let mut wire = std::vec![0, 0];
        wire.extend_from_slice(options);
        wire
    }

    #[test]
    fn multicast_without_predicates_resets_trickle() {
        let mut timer = timer();
        assert_eq!(timer.interval, 8_000);
        assert_eq!(timer.counter, 1);

        let action =
            handle_authenticated_dis(&[0xff, 0], true, CONTEXT, &mut timer, 10_000, 0).unwrap();

        assert_eq!(action, DisAction::ResetTrickle);
        assert_eq!(timer.interval, 4_000);
        assert_eq!(timer.interval_start, 10_000);
        assert_eq!(timer.counter, 0);
    }

    #[test]
    fn every_matching_predicate_combination_resets() {
        for flags in [0x00, 0x20, 0x40, 0x80, 0xe0, 0xff] {
            let mut timer = timer();
            let wire = dis_with_options(&solicited(flags, 7, CONTEXT.dodag_id, 9));
            assert_eq!(
                handle_authenticated_dis(&wire, true, CONTEXT, &mut timer, 10_000, 1),
                Ok(DisAction::ResetTrickle)
            );
        }
    }

    #[test]
    fn enabled_predicate_mismatch_is_ignored_without_mutation() {
        let cases = [
            solicited(INSTANCE_PREDICATE, 8, CONTEXT.dodag_id, 9),
            solicited(DODAG_PREDICATE, 7, [0x33; 16], 9),
            solicited(VERSION_PREDICATE, 7, CONTEXT.dodag_id, 10),
        ];
        for option in cases {
            let mut timer = timer();
            let before = (
                timer.interval,
                timer.interval_start,
                timer.counter,
                timer.transmit_time,
            );
            let wire = dis_with_options(&option);
            assert_eq!(
                handle_authenticated_dis(&wire, true, CONTEXT, &mut timer, 10_000, 0),
                Ok(DisAction::Ignore)
            );
            assert_eq!(
                (
                    timer.interval,
                    timer.interval_start,
                    timer.counter,
                    timer.transmit_time,
                ),
                before
            );
        }
    }

    #[test]
    fn disabled_predicates_and_reserved_flags_are_ignored() {
        let mut timer = timer();
        let wire = dis_with_options(&solicited(0x1f, 255, [0x33; 16], 255));
        assert_eq!(
            handle_authenticated_dis(&wire, true, CONTEXT, &mut timer, 10_000, 0),
            Ok(DisAction::ResetTrickle)
        );
    }

    #[test]
    fn unicast_match_requests_configured_dio_without_reset() {
        for wire in [
            std::vec![0, 0],
            dis_with_options(&solicited(0xe0, 7, CONTEXT.dodag_id, 9)),
        ] {
            let mut timer = timer();
            let before = (
                timer.interval,
                timer.interval_start,
                timer.counter,
                timer.transmit_time,
            );
            assert_eq!(
                handle_authenticated_dis(&wire, false, CONTEXT, &mut timer, 10_000, 0),
                Ok(DisAction::UnicastDioWithConfiguration)
            );
            assert_eq!(
                (
                    timer.interval,
                    timer.interval_start,
                    timer.counter,
                    timer.transmit_time,
                ),
                before
            );
        }
    }

    #[test]
    fn unicast_predicate_mismatch_is_ignored() {
        let mut timer = timer();
        let wire = dis_with_options(&solicited(VERSION_PREDICATE, 7, CONTEXT.dodag_id, 8));
        assert_eq!(
            handle_authenticated_dis(&wire, false, CONTEXT, &mut timer, 10_000, 0),
            Ok(DisAction::Ignore)
        );
    }

    #[test]
    fn malformed_or_duplicate_options_fail_before_timer_mutation() {
        let malformed = [
            std::vec![0],
            std::vec![0, 1],
            std::vec![0, 0, OPT_SOLICITED_INFORMATION],
            dis_with_options(&[OPT_SOLICITED_INFORMATION, 19, 0]),
            dis_with_options(&[
                OPT_SOLICITED_INFORMATION,
                18,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
            ]),
            {
                let option = solicited(0, 0, [0; 16], 0);
                let mut options = option.to_vec();
                options.extend_from_slice(&option);
                dis_with_options(&options)
            },
        ];
        for wire in malformed {
            let mut timer = timer();
            let before = (
                timer.interval,
                timer.interval_start,
                timer.counter,
                timer.transmit_time,
            );
            assert!(handle_authenticated_dis(&wire, true, CONTEXT, &mut timer, 10_000, 0).is_err());
            assert_eq!(
                (
                    timer.interval,
                    timer.interval_start,
                    timer.counter,
                    timer.transmit_time,
                ),
                before
            );
        }
    }

    #[test]
    fn padding_and_unknown_options_do_not_hide_valid_solicitation() {
        let mut options = std::vec![0, 0xee, 1, 0xaa];
        options.extend_from_slice(&solicited(0xe0, 7, CONTEXT.dodag_id, 9));
        let wire = dis_with_options(&options);
        let mut timer = timer();
        assert_eq!(
            handle_authenticated_dis(&wire, true, CONTEXT, &mut timer, 10_000, 0),
            Ok(DisAction::ResetTrickle)
        );
    }
}
